appsec/src/helper/remote_config/client.cpp

```diff
@@ -6,240 +6,186 @@
+        throw std::runtime_error{
+            "Failed to resolve ddog_remote_config_reader_for_path"};
+    }
 
-    std::smatch base_match;
-    if (!std::regex_match(path, base_match, regex) || base_match.size() < 4) {
-        throw invalid_path();
+    ddog_remote_config_read =
+        // NOLINTNEXTLINE
+        reinterpret_cast<decltype(ddog_remote_config_read)>(
+            dlsym(RTLD_DEFAULT, "ddog_remote_config_read"));
+    if (ddog_remote_config_read == nullptr) {
+        throw std::runtime_error{"Failed to resolve ddog_remote_config_read"};
     }
 
-    return config_path{base_match[3].str(), base_match[2].str()};
+    ddog_remote_config_reader_drop =
+        // NOLINTNEXTLINE
+        reinterpret_cast<decltype(ddog_remote_config_reader_drop)>(
+            dlsym(RTLD_DEFAULT, "ddog_remote_config_reader_drop"));
+    if (ddog_remote_config_reader_drop == nullptr) {
+        throw std::runtime_error{
+            "Failed to resolve ddog_remote_config_reader_drop"};
+    }
 }
 
-client::client(std::unique_ptr<http_api> &&arg_api, service_identifier &&sid,
-    remote_config::settings settings,
-    std::vector<listener_base::shared_ptr> listeners)
-    : api_(std::move(arg_api)), id_(dds::generate_random_uuid()),
-      sid_(std::move(sid)), settings_(std::move(settings)),
-      listeners_(std::move(listeners))
+client::client(remote_config::settings settings,
+    std::vector<std::shared_ptr<listener_base>> listeners)
+    : reader_{ddog_remote_config_reader_for_path(settings.shmem_path.c_str()),
+          ddog_remote_config_reader_drop},
+      settings_{std::move(settings)}, listeners_{std::move(listeners)}
 {
+    assert(settings.enabled == true); // NOLINT
+
     for (auto const &listener : listeners_) {
-        const auto &supported_products = listener->get_supported_products();
-        for (const auto &[name, capabilities] : supported_products) {
-            products_.insert(std::pair<std::string, remote_config::product>(
-                name, product(name, listener)));
-            capabilities_ |= capabilities;
+        for (const product p : listener->get_supported_products()) {
+            std::vector<listener_base *> &vec_listeners =
+                listeners_per_product_[p];
+            vec_listeners.push_back(listener.get());
         }
     }
 }
 
-client::ptr client::from_settings(service_identifier &&sid,
+std::unique_ptr<client> client::from_settings(
     const remote_config::settings &settings,
-    std::vector<listener_base::shared_ptr> listeners)
+    std::vector<std::shared_ptr<listener_base>> listeners)
 {
-    return std::make_unique<client>(std::make_unique<http_api>(settings.host,
-                                        std::to_string(settings.port)),
-        std::move(sid), settings, std::move(listeners));
+    return std::unique_ptr<client>{new client(settings, std::move(listeners))};
 }
 
-[[nodiscard]] protocol::get_configs_request client::generate_request() const
+bool client::poll()
```

Comment:
Maybe I missed it but is there any tests on helper tests checking this? I saw the client_test got removed

Reply:
You're right, I removed it. These are tested only on the integration tests. Reading of remote config is done by calling `ddog_remote_config_read`, which is a function implemented in Rust that we import.

Any test using the real Rust read function would writing in shared memory in the format expected by that rust code. By that time, we're having the test depend on implementation details of the rust code.

And by mocking `ddog_remote_config_read`, we're testing much useful stuff anymore.